crates/uv-client/src/registry_client.rs

```diff
@@ -247,7 +247,21 @@ impl RegistryClient {
             async {
                 // Use the response URL, rather than the request URL, as the base for relative URLs.
                 // This ensures that we handle redirects and other URL transformations correctly.
-                let url = response.url().clone();
+                let request_url = url;
+                let mut url = response.url().clone();
+
+                // We must ensure authentication is retained for hosts
+                if request_url.host() == url.host() {
```

Comment:
We should check the protocol too, to avoid potential downgrades. 


This seems uhm, dangerous

Reply:
Should be resolved by https://github.com/astral-sh/uv/pull/1874/commits/c57ef0b4fe3744965c7222a290bb49b775e11a60